### PR TITLE
Implement correct division by zero behavior

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description

Implement correct behavior for division and modulo by zero.

## Testing

Existing tests.

## Documentation

No.

Resolves: #1410

### CODE REVIEW

File: tools/bpf2c/bpf_code_generator.cpp [hunk 1]
```diff
@@ -557,17 +548,12 @@ bpf_code_generator::encode_instructions(const bpf_code_generator::unsafe_string&
...
+                output.lines.push_back(format_string("if (%s != 0) {", source));
                 if (is64bit)
-                    output.lines.push_back(format_string("%s %%= %s;", destination, source));
+                    output.lines.push_back(format_string(INDENT "%s %%= %s;\n" INDENT "}", destination, source));
                 else
-                    output.lines.push_back(
-                        format_string("%s = (uint32_t)%s %% (uint32_t)%s;", destination, destination, source));
+                    output.lines.push_back(format_string(
+                        INDENT "%s = (uint32_t)%s %% (uint32_t)%s;\n" INDENT "}", destination, destination, source));
                 break;
```
Comment (dthaler) [inline tools/bpf2c/bpf_code_generator.cpp:557]: Where is the else case where source == 0?
I see the else clause in 525-530 for division, but I don't see the equivalent here for modulo.

File: tools/bpf2c/bpf_code_generator.cpp [hunk 2]
```diff
@@ -557,17 +551,17 @@ bpf_code_generator::encode_instructions(const bpf_code_generator::unsafe_string&
...
-                if (inst.opcode & EBPF_SRC_REG) {
-                    output.lines.push_back(check_div_by_zero);
-                } else if (inst.imm == 0) {
-                    throw bpf_code_generator_exception(
-                        "invalid instruction - constant division by zero", output.instruction_offset);
-                }
                 if (is64bit)
-                    output.lines.push_back(format_string("%s %%= %s;", destination, source));
+                    output.lines.push_back(format_string(
+                        "%s = %s ? (%s %% %s): %s ;", destination, source, destination, source, destination));
```
Comment (dthaler) [inline tools/bpf2c/bpf_code_generator.cpp:556]: I think this is a bug and ought to be
```suggestion
                        "%s = %s ? (%s %% %s): %s ;", destination, source, destination, source, source));
```
Same with line 564

Comment (Alan-Jowett) [reply tools/bpf2c/bpf_code_generator.cpp:556]: Assume destination = r1 and source = r2.

The instruction was:
r1 %= r2

After substitution

Current change:
r1 = r2 ? (r1 % r2) : r1;

Proposed change:
r1 = r2 ? (r1 % r2) : r2;

If (r2 == 0) then this becomes (r1 = 0) which doesn't match what I thought the required behavior was.

Comment (Alan-Jowett) [reply tools/bpf2c/bpf_code_generator.cpp:556]: Assume destination = r1 and source = r2.

The instruction was:
r1 %= r2

After substitution

Current change:
r1 = r2 ? (r1 % r2) : r1;

Proposed change:
r1 = r2 ? (r1 % r2) : r2;

If (r2 == 0) then this becomes (r1 = 0) which doesn't match what I thought the required behavior was.

Comment (dthaler) [reply tools/bpf2c/bpf_code_generator.cpp:556]: My bad, the original code was right and my doc was wrong.  It should be set to dest (r1) as current

**Review by dthaler — DISMISSED** [review]

File: tests/bpf2c_tests/expected/divide_by_zero_dll.c [hunk 3]
```diff
@@ -154,12 +154,7 @@ divide_by_zero(void* context)
     r6 = IMMEDIATE(100000);
     // EBPF_OP_DIV64_REG pc=10 dst=r6 src=r1 offset=0 imm=0
 #line 32 "sample/divide_by_zero.c"
-    if (r1 == 0) {
-        division_by_zero(10);
-        return 0xffffffffffffffffui64;
-    }
-#line 32 "sample/divide_by_zero.c"
-    r6 /= r1;
+    r6 = r1 ? (r6 / r1) : 0;
```
Comment (saxena-anurag) [inline tests/bpf2c_tests/expected/divide_by_zero_dll.c:157]: Should we add a sample program for mod also?

**Review by dthaler — APPROVED** [review]

File: tests/bpf2c_tests/expected/divide_by_zero_dll.c (hunk 3 above)
Comment (Alan-Jowett) [reply tests/bpf2c_tests/expected/divide_by_zero_dll.c:157]: That path is covered by the tests we pull from uBPF and then run through bpf2c.

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**